### PR TITLE
[hail] Interpret uses `LoweringPipeline`

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by iitalics** [comment]
looked into the failing tests concerning randomness, and tracked down the source of the failures:

- `ApplySeeded` is an `AbstractApplyNode`. this means the interpreter will try to "memoize" the function definition so as to generate it only once even if you call it in multiple places in the IR. 
- memoization is based on referential equality, so they need to be the exact same IR object in order for reuse to trigger
- the "seeded function" implementations used by the test suite will create a new randomness state per generated function
- the Interpret pipeline used to rewrite each ApplySeeded node to be different objects, but with your changes they are the same, so their functions get memoized, thus sharing the same state

im not entirely sure which of these was doing the wrong thing, but IMO your changes in this PR should not have introduced any problems

**Comment by tpoterba** [comment]
Thanks for looking into this, Milo!

**Comment by tpoterba** [comment]
Yeah, this PR really shouldn't have introduced the kinds of failures we're seeing. That seems to be the refrain of all of my lowering refactoring PRs though!

File: hail/src/test/scala/is/hail/expr/ir/RandomFunctionsSuite.scala [hunk 1]
```diff
@@ -51,7 +51,7 @@ class RandomFunctionsSuite extends HailSuite {
 
   implicit val execStrats = ExecStrategy.javaOnly
 
-  val counter = ApplySeeded("counter_seeded", FastSeq(), 0L, TInt32())
+  def counter = ApplySeeded("counter_seeded", FastSeq(), 0L, TInt32())
```
Comment (tpoterba) [inline hail/src/test/scala/is/hail/expr/ir/RandomFunctionsSuite.scala]: @catoverdrive This fixes the bug. The problem is that Interpret's function memoization is doing the wrong thing when seeded function IRs appear with the same reference in the IR tree. We could deep copy going into interpret, I suppose?

**Comment by tpoterba** [comment]
should pass now!

**Pull request merged**